packages/vats/tools/ibc-utils.js

```diff
@@ -0,0 +1,84 @@
+export const REMOTE_ADDR_RE =
+  /^(\/ibc-hop\/[^/]+)*\/ibc-port\/([^/]+)\/(ordered|unordered)\/([^/]+)$/s;
+harden(REMOTE_ADDR_RE);
+/** @typedef {`/ibc-hop/${string}/ibc-port/${string}/{'ordered'|'unordered'}/${string}`} RemoteIbcAddress */
+
+export const LOCAL_ADDR_RE = /^\/ibc-port\/([-a-zA-Z0-9._+#[\]<>]+)$/;
+harden(LOCAL_ADDR_RE);
+/** @typedef {`/ibc-port/${string}`} LocalIbcAddress */
+
+/** @param {string} remoteAddr */
+export const decodeRemoteIbcAddress = remoteAddr => {
+  const match = remoteAddr.match(REMOTE_ADDR_RE);
+  if (!match) {
+    throw TypeError(
+      `Remote address ${remoteAddr} must be '(/ibc-hop/CONNECTION)*/ibc-port/PORT/(ordered|unordered)/VERSION'`,
+    );
+  }
+
+  const hops = [];
+  let h = match[1];
+  while (h) {
+    const m = h.match(/^\/ibc-hop\/([^/]+)/);
+    if (!m) {
+      throw Error(
+        `internal: ${JSON.stringify(h)} did not begin with "/ibc-hop/XXX"`,
+      );
+    }
+    h = h.substr(m[0].length);
+    hops.push(m[1]);
+  }
+
+  // Generate a circuit.
+  const rPortID = match[2];
+  const order =
+    match[3] === 'ordered'
+      ? /** @type {const} */ ('ORDERED')
+      : /** @type {const} */ ('UNORDERED');
+  const version = match[4];
+  return { rPortID, hops, order, version };
+};
+harden(decodeRemoteIbcAddress);
+
+/**
+ * @param {LocalIbcAddress} localAddr
+ * @returns {string}
+ */
+export const localAddrToPortID = localAddr => {
+  const m = localAddr.match(LOCAL_ADDR_RE);
+  if (!m) {
+    throw TypeError(
+      `Invalid port specification ${localAddr}; expected "/ibc-port/PORT"`,
+    );
+  }
+  return m[1];
+};
+harden(localAddrToPortID);
+
+/**
+ * @param {string[]} hops
+ * @param {string} rPortID
+ * @param {'ordered' | 'unordered' | 'ORDERED' | 'UNORDERED'} order
+ * @param {string} rVersion
+ * @param {string} rChannelID
+ * @returns {RemoteIbcAddress}
+ */
+export const encodeRemoteIbcAddress = (
+  hops,
+  rPortID,
+  order,
+  rVersion,
+  rChannelID,
+) => {
+  const ibcHops = hops.map(hop => `/ibc-hop/${hop}`).join('/');
+  return /** @type {RemoteIbcAddress} */ (
+    `${ibcHops}/ibc-port/${rPortID}/${order.toLowerCase()}/${rVersion}/ibc-channel/${rChannelID}`
+  );
+};
+harden(encodeRemoteIbcAddress);
+
+export const encodeLocalIbcAddress = (portID, order, version) => {
```

Comment:
I think we might be able to add types for `portID` and `rPortID`. Maybe something like: 
```ts
type LocalPortID =  `icacontroller-${number}` | `icqcontroller-${number}`;
type RemotePortID = 'transfer' | 'icahost' | 'icqhost';

```

Reply:
I disagree here.  That would be overspecification, which is a pitfall that ibc-go made when first specifying ICA.  The names of ports (_Identifiers_) should be as general [as what the IBC spec allows](https://github.com/cosmos/ibc/blob/00ee5a27819570f871e20d01630e4c2eb19b3a20/spec/core/ics-024-host-requirements/README.md#paths-identifiers-separators), not a restricted set of "what we think people will use".

Reply:
Good feedback, thank you. Maybe these definitions should instead live in `orchestration` and extend the more generalized specification in here?